### PR TITLE
Add MSI garbage collection for side-by-side workload manifests

### PR DESCRIPTION
**Pull request by dsplaisted**
- Add MSI garbage collection for side-by-side workload manifests
- Remove MSI upgrade logic since all workload MSIs are now side by side
- Current SxS manifests don't write installation records (fixed with https://github.com/dotnet/arcade/pull/14359), so this PR includes fallback logic which finds the manifests on disk and then looks up the installation information in the registry
- Change file-based workload pack installation records to be based on the resolved pack ID for aliased packs, to be consistent with how MSIs work (and work better with the workload garbage collector).  This means there are some SxS scenarios for file-based installs where aliased workload packs would not be garbage collected correctly.

### CODE REVIEW

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+using Microsoft.Win32.Msi;
+using NuGet.Packaging.Core;
+using NuGet.Versioning;
+using static Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadResolver;
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    internal partial class NetSdkMsiInstallerClient
+    {
+        internal static readonly Guid UpgradeCodeNamespaceUuid = Guid.Parse("C743F81B-B3B5-4E77-9F6D-474EFF3A722C");
```
Comment (joeloff) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:21]: We should probably add a comment here to link this back to the Arcade namespace GUID

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+                                continue;
+                            }
+
+                            Log.LogMessage($"Discovered manifest installation in {manifestVersionFolder} which didn't have corresponding installation record in Registry");
+
+                            if (NuGetVersion.TryParse(manifestVersionString, out NuGetVersion manifestVersion))
+                            {
+                                var packageIdentity = new PackageIdentity(manifestID + ".Manifest-" + manifestFeatureBand, manifestVersion);
+                                string uuidName = $"{packageIdentity};{HostArchitecture}";
+                                var upgradeCode = '{' + CreateUuid(UpgradeCodeNamespaceUuid, uuidName).ToString() + '}';
```
Comment (joeloff) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:135]: Can probably simplify this as `var upgradeCode = $"{{{CreateUuid(UpgradeCodeNamespaceUuid, uuidName)}}}";`

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+            //  Workload manifest MSIs for 8.0.100 don't yet write the same type of installation records to the registry that workload packs do.
+            //  So to find what is installed, we look for the manifests on disk, and then map that to installed MSIs
+            //  To do the mapping to installed MSIs, we rely on the fact that the MSI UpgradeCode is generated in a stable fashion from the
+            //  NuGet package identity and platform: Utils.CreateUuid(UpgradeCodeNamespaceUuid, $"{Package.Identity};{Platform}")
+            //  The NuGet package identity used is the vanilla (non-MSI) manifest package, for example Microsoft.NET.Workload.Mono.ToolChain.Current.Manifest-8.0.100 version 8.0.0
+
+            //  Manifest IDs are lowercased on disk, we need to map them back to the original casing to generate the right UpgradeCode
+            Dictionary<string, string> manifestIDCasing = new Dictionary<string, string>();
+            foreach (var casedManifestID in new[]
```
Comment (joeloff) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:78]: should we move the array to a static readonly private field and avoid allocating it each time the method runs?

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+                                    manifestRecord.ProductVersion = installedVersion;
+                                    manifestRecord.ProviderKeyName = dependencyProvider.ProviderKeyName;
+
+                                    manifestRecords.Add(manifestRecord);
+
+                                    Log.LogMessage($"Found installed manifest: {manifestRecord.ProviderKeyName}, {manifestRecord.ProductCode}");
+                                }
+                                else if (relatedProductCodes.Count > 1)
+                                {
+                                    Log.LogMessage($"Found multiple product codes for {uuidName}, which is not expected for side-by-side manifests.");
```
Comment (joeloff) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:167]: For this case, I think we should consider logging the ProductCode values to assist with troubleshooting.

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs [hunk 5]
```diff
@@ -108,38 +128,64 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
...
-                foreach (var expectedPack in installedPackInfos)
+                List<WorkloadManifestRecord> manifestsToRemove = new();
+
+                var installedWorkloadManifests = GetWorkloadManifestRecords();
+                foreach (var manifestRecord in installedWorkloadManifests)
                 {
-                    if (!expectedPack.Id.ToString().Equals(expectedPack.ResolvedPackageId, StringComparison.OrdinalIgnoreCase))
+                    DependencyProvider depProvider = new DependencyProvider(manifestRecord.ProviderKeyName);
+
+                    (bool shouldBeInstalled, string reason) shouldBeInstalled(SdkFeatureBand dependentFeatureBand)
```
Comment (MiYanni) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:144]: It seems that we (were) recommending using PascalCase for local functions. https://stackoverflow.com/a/75593010/294804

Not sure that is still the case but I've generally treated them as methods, so I've used PascalCase myself.

**Review by MiYanni — APPROVED** [review]

**Review by joeloff — APPROVED** [review]
Looks good, thank you.

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs (hunk 2 above)
Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:135]: To me that's not actually simpler, with all the repeated braces, two to escape a brace and another one for the string interpolation.

File: src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+                "Microsoft.NET.Sdk.MacCatalyst",
+                "Microsoft.NET.Sdk.macOS",
+                "Microsoft.NET.Sdk.Maui",
+                "Microsoft.NET.Sdk.tvOS",
+                "Microsoft.NET.Workload.Emscripten.Current",
+                "Microsoft.NET.Workload.Emscripten.net6",
+                "Microsoft.NET.Workload.Emscripten.net7",
+                "Microsoft.NET.Workload.Mono.ToolChain.Current",
+                "Microsoft.NET.Workload.Mono.ToolChain.net6",
+                "Microsoft.NET.Workload.Mono.ToolChain.net7",
```
Comment (akoeplinger) [inline src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:36]: @dsplaisted I assume this list will need to be updated to add the `.net8` packages in main right?

Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:36]: @akoeplinger This list of IDs is actually part of a workaround for the fact that the manifest MSIs weren't writing a manifest install record to the registry.  We added that in https://github.com/dotnet/arcade/pull/14359, so the .net8 manifest MSIs (which are produced in main) should have the right installation records and shouldn't need this fallback logic to be detected.

Comment (akoeplinger) [reply src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs:36]: Perfect, thanks

**Pull request merged**